README.rst
=================
Flake8 Spellcheck
=================

|CircleCI| |Black| |PyPi|

Flake8 Plugin that spellchecks variables, functions, classes and other bits of your python code.

You can whitelist words that are specific to your project simply by adding them to ``spellcheck-allowlist.txt``
in the root of your project directory. Each word you add should be separated by a newline.

Spelling is assumed to be in en_US.

This plugin supports python 3.8+

Codes
=====

* SC100 - Spelling error in comments
* SC200 - Spelling error in name (e.g. variable, function, class)

Configuration
=============

Enable non-default dictionaries
-------------------------------

This plugin comes with a couple of dictionary files that aren't enabled by default.

- ``django``
- ``pytest``

To enable a custom dictionary, you can use the ``--spellcheck-add-dictionary`` flag.

.. code-block::

    flake8 --spellcheck-add-dictionary django --spellcheck-add-dictionary pytest mypackage

Disable default dictionaries
----------------------------

The default dictionary list is as follows:

- ``en_US``
- ``technical``
- ``python``

You can disable the default dictionary list, if you want to be highly specific about the wordlist in use.

.. code-block::

    flake8 --spellcheck-disable-default-dictionaries mypackage

Or in your flake8 configuration (e.g. your ``.flake8`` file):

.. code-block:: ini

    spellcheck-disable-default-dictionaries = true

You can then enable built-in dictionaries one-by-one as necessary.

.. code-block::

    flake8 --spellcheck-disable-default-dictionaries --spellcheck-add-dictionary python --spellcheck-add-dictionary pytest --spellcheck-add-dictionary pandas mypackage

Or in your flake8 configuration file:

.. code-block:: ini

    spellcheck-add-dictionary = python,pytest,pandas

Specify Code Targets
--------------------

Both ``comments`` and ``names`` (variable names, function names...) are spellchecked by default.
You can specify what targets to spellcheck by specifying the ``--spellcheck-targets`` flag.

.. code-block::

    flake8 --spellcheck-targets comments mypackage

Or by setting the ``spellcheck-targets`` option in your flake8 configuration (e.g. in your ``.flake8`` file).

.. code-block:: ini

   [flake8]
   spellcheck-targets=comments

The above configuration would only spellcheck comments, while this configuration will only check symbol names.

.. code-block:: ini

   [flake8]
   spellcheck-targets=names

Contributing
============

If you have found word(s) which are listed as a spelling error but are actually correct terms used
in python or in technical implementations (e.g. http), then you can very easily contribute by
adding those word(s) to the appropriate dictionaries:

* `python dictionary <flake8_spellcheck/python.txt>`_
* `pytest dictionary <flake8_spellcheck/pytest.txt>`_
* `general technical dictionary <flake8_spellcheck/technical.txt>`_
* `django dictionary <flake8_spellcheck/django.txt>`_
* `pandas dictionary <flake8_spellcheck/pandas.txt>`_

Before you submit a PR, it is recommended to run ``check-sorting.sh`` in the root of this repository,
to verify that all the dictionary files are still sorted correctly. Sorting is enforced by CI, so
you'll need to make sure the files are sorted before your PR can be merged.

Development
-----------

* Install `poetry <https://github.com/python-poetry>`__
* Run ``poetry install``
* Run ``poetry run pre-commit install --install-hooks``

You can run tests with ``poetry run pytest``.


.. |CircleCI| image:: https://circleci.com/gh/MichaelAquilina/flake8-spellcheck.svg?style=svg
   :target: https://circleci.com/gh/MichaelAquilina/flake8-spellcheck

.. |PyPi| image:: https://badge.fury.io/py/flake8-spellcheck.svg
   :target: https://badge.fury.io/py/flake8-spellcheck

.. |Black| image:: https://img.shields.io/badge/code%20style-black-000000.svg
   :target: https://github.com/psf/black


CHANGELOG.md
Flake8 Spellcheck Changelog
===========================

0.28.0
------
* Use poetry as a build backend
* Skipped a minor version due to git commit mistake

0.26.0
------
* Drop support for python 3.6
* Remove dependency on importlib
* General codebase improvements (Thanks @shaleh and @djmattyg007)
* Add type annotations
* Introduce pre-commit and mypy to codebase

0.25.0
------
* Add partial support for pandas dictionaries (Thanks @aaronzipp)
* Update python dictionary with missing words

0.24.0
------
* Speed up run for multiple files by loading once the dictionary, and using ``importlib.resources`` instead of
  ``pkg_resources`` - by @gaborbernat.

0.21.0 - 0.23.0
---------------
* Add more common words to technical and python dictionaries

0.20.1
------
* Fix crash on lines with only '#' characters (Thanks @paulovcmedeiros)

0.20.0
------
* Updated techincal and django dictionaries (Thanks @3n-k1)

0.19.1
------
* Fix bug where # noqa comments were incorrectly spellchecked if combined with typing comments (Thanks @3n-k1 for the fix)

0.19.0
------
* Updates to technical and python dictionaries

0.18.1
------
* Fix regression where plugin would crash on lines with empty comments (Issue #34)

0.18.0
------
* Do not report errors on `# noqa:` comments. Thanks @3n-k1!

0.17.0
------
* Update technical.txt

0.16.0
------
* Add support for specifying spellcheck targets in flake8 configuration

0.15.0
------
* Update technical and en_US

0.14.0
------
* Add "hardcode" to technical.txt

0.13.0
------
* Add more words to python, django and technical dictionaries (Thanks @pbelskiy)

0.12.1
------
* Fix bug where quotes in comments would result in spell errors

0.12.0
------
* Add more words to python dictionary
* Fully support optional django dictionary

0.11.0
------
* Add more words to python dictionary
* Add initial support for optional in-built dictionaries (e.g. Django)

0.10.0
------
* Handle special cases when spellchecking words ending with 's

0.9.1
-----
* Add `hasattr` to in-built python dictionary

0.9.0
-----
* Introduce Changelog
* Correctly detect fully UPPERCASE variables (thanks @gera)
* Correctly detect methods and variables with leading underscores (thanks @gera)
* Added new inbuilt words to python dictionary


tests/test_flake8_spellcheck.py
from textwrap import dedent

import pytest

from flake8_spellcheck import is_number, parse_camel_case, parse_snake_case


@pytest.mark.parametrize(
    ["value", "col_offset", "tokens"],
    [
        ("bad_function", (0, 0), [((0, 0), "bad"), ((0, 4), "function")]),
        ("`pre_written`", (1, 2), [((1, 3), "pre"), ((1, 7), "written")]),
        (
            "foo_bar_baz",
            (30, 4),
            [((30, 4), "foo"), ((30, 8), "bar"), ((30, 12), "baz")],
        ),
        ("__init__", (0, 3), [((0, 5), "init")]),
    ],
)
def test_parse_snake_case(value, col_offset, tokens):
    assert list(parse_snake_case(value, col_offset)) == tokens


@pytest.mark.parametrize(
    ["value", "col_offset", "tokens"],
    [
        ("FakeClass", (0, 0), [((0, 0), "Fake"), ((0, 4), "Class")]),
        ("don't", (0, 0), [((0, 0), "don't")]),
        ("coding:", (20, 10), [((20, 10), "coding")]),
        ("`FastCar`", (1, 22), [((1, 23), "Fast"), ((1, 27), "Car")]),
        ("pair-programming", (5, 0), [((5, 0), "pair"), ((5, 5), "programming")]),
        ("FooBarBaz", (4, 4), [((4, 4), "Foo"), ((4, 7), "Bar"), ((4, 10), "Baz")]),
        ("_ignoredValue", (20, 10), [((20, 11), "ignored"), ((20, 18), "Value")]),
    ],
)
def test_parse_camel_case(value, col_offset, tokens):
    assert list(parse_camel_case(value, col_offset)) == tokens


@pytest.mark.parametrize(["value", "result"], [("8", True), ("word8", False)])
def test_is_number(value, result):
    assert is_number(value) is result


def test_python_words(flake8_path):
    (flake8_path / "example.py").write_text(
        dedent(
            """
            id = str(4)
            if isinstance(id, int):
                dict(id=id)
            """
        )
    )
    result = flake8_path.run_flake8()
    assert result.out_lines == []


class TestComments:
    def test_fail(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                # dont "make" b4d c8omm3nts
                foo = "bar"
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.exit_code == 1
        assert result.out_lines == [
            "./example.py:2:1: SC100 Possibly misspelt word: 'dont'",
            "./example.py:2:1: SC100 Possibly misspelt word: 'b4d'",
            "./example.py:2:1: SC100 Possibly misspelt word: 'c8omm3nts'",
        ]

    def test_pass(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                # don't "make" 'bad' comments
                foo = "bar"
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.out_lines == []

    def test_disabled(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                # dont "make" b4d c8omm3nts
                foo = "bar"
                """
            )
        )
        result = flake8_path.run_flake8(["--spellcheck-targets=names"])
        assert result.exit_code == 0
        assert result.out_lines == []

    def test_flake8_pragma(self, flake8_path):
        (flake8_path / "example.py").write_text("foo = 'bar'  # noqa: W503\n")
        result = flake8_path.run_flake8()
        assert result.out_lines == []

    def test_flake8_pragma_spaces(self, flake8_path):
        (flake8_path / "example.py").write_text("foo = 'bar'  #    noqa: W503\n")
        result = flake8_path.run_flake8()
        assert result.out_lines == [
            "./example.py:1:14: E262 inline comment should start with '# '"
        ]

    # Regression test for github.com/MichaelAquilina/flake8-spellcheck/issues/34
    def test_empty_comment(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                # the empty comment below should not fail
                #
                # hello world
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.out_lines == []

    # Regression test for https://github.com/MichaelAquilina/flake8-spellcheck/issues/36
    def test_type_and_noqa(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                # the comment below should not fail on `W503` or `W504`  # noqa: SC100
                foo = []  # type: ignore  # noqa: W503  # noqa: W504
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.out_lines == []

    # Regression test for github.com/MichaelAquilina/flake8-spellcheck/issues/40
    def test_pure_number_char_comment(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                # the comment below should not fail
                ##########
                # hello world
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.out_lines == []


class TestFunctionDef:
    def test_apostrophe(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                def dont_fail(a):
                    return a + 2


                def cant_fail(b):
                    return b * 4
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.exit_code == 0
        assert result.out_lines == []

    def test_apostrophe_ending_with_s(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                def request_classs(a, b, c):
                    pass
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.exit_code == 1
        assert result.out_lines == ["./example.py:2:13: SC200 Possibly misspelt word: 'classs'"]

    def test_fail(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                def mispleled_function(a, b, c):
                    pass
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.exit_code == 1
        assert result.out_lines == ["./example.py:2:5: SC200 Possibly misspelt word: 'mispleled'"]

    def test_pass(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                def misspelled_function(a, b, c):
                    pass
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.exit_code == 0
        assert result.out_lines == []

    def test_disabled(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                def mispleled_function(a, b, c):
                    pass
                """
            )
        )
        result = flake8_path.run_flake8(["--spellcheck-targets=comments"])
        assert result.exit_code == 0
        assert result.out_lines == []


class TestName:
    def test_fail(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                my_varaible_namde = "something"
                SOMETHIGN = "SOMETHING"
                SOMETHING_ELS = "SOMETHING"
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.exit_code == 1
        assert result.out_lines == [
            "./example.py:2:4: SC200 Possibly misspelt word: 'varaible'",
            "./example.py:2:13: SC200 Possibly misspelt word: 'namde'",
            "./example.py:3:1: SC200 Possibly misspelt word: 'SOMETHIGN'",
            "./example.py:4:11: SC200 Possibly misspelt word: 'ELS'",
        ]

    def test_pass(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                my_variable_name = "something"
                SOMETHING = "SOMETHING"
                SOMETHING_ELSE = "SOMETHING"
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.exit_code == 0
        assert result.out_lines == []

    def test_disabled(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                my_varaible_namde = "something"
                SOMETHIGN = "SOMETHING"
                SOMETHING_ELS = "SOMETHING"
                """
            )
        )
        result = flake8_path.run_flake8(["--spellcheck-targets=comments"])
        assert result.exit_code == 0
        assert result.out_lines == []


class TestClassDef:
    def test_fail(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                class FackeClaessName:
                    pass
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.exit_code == 1
        assert result.out_lines == [
            "./example.py:2:7: SC200 Possibly misspelt word: 'Facke'",
            "./example.py:2:12: SC200 Possibly misspelt word: 'Claess'",
        ]

    def test_pass(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                class FakeClassName:
                    pass
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.out_lines == []

    def test_disabled(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                class FackeClaessName:
                    pass
                """
            )
        )
        result = flake8_path.run_flake8(["--spellcheck-targets=comments"])
        assert result.exit_code == 0
        assert result.out_lines == []


class TestLeadingUnderscore:
    def test_fail(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                def _doSsomething(s):
                    pass
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.exit_code == 1
        assert result.out_lines == ["./example.py:2:8: SC200 Possibly misspelt word: 'Ssomething'"]

    def test_pass(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                def _doSomething(s):
                    pass
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.exit_code == 0
        assert result.out_lines == []


class TestOptionalDictionaries:
    def test_fail(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                from django.http import HttpResponse
                from django.views.decorators.csrf import csrf_protect


                @csrf_protect
                def my_view(request):
                    return HttpResponse("hello world")
                """
            )
        )
        result = flake8_path.run_flake8()
        assert result.exit_code == 1
        assert result.out_lines == [
            "./example.py:3:30: SC200 Possibly misspelt word: 'csrf'",
            "./example.py:3:42: SC200 Possibly misspelt word: 'csrf'",
            "./example.py:6:2: SC200 Possibly misspelt word: 'csrf'",
        ]

    def test_success(self, flake8_path):
        (flake8_path / "example.py").write_text(
            dedent(
                """
                from django.http import HttpResponse
                from django.views.decorators.csrf import csrf_protect


                @csrf_protect
                def my_view(request):
                    return HttpResponse("hello world")
                """
            )
        )
        result = flake8_path.run_flake8(["--dictionaries=python,technical,django,en_US"])
        assert result.exit_code == 0
        assert result.out_lines == []


whitelist.py
import re


def write_Whitelist():
    """
    add words to the whitelist.txt for spelling check
    """
    words = dict()
    whitelist = open(r'/Users/logan.li/project/django.txt', 'a')

    # cache: run `flake8 django/ > whitelist.cache`
    with open(r'/Users/logan.li/project/whitelist.cache', 'r+') as cache:
        for line in cache.readlines():
            word = re.match(".* SC200 .*'(.*)'", line)
            if word:
                word = word.group(1).lower()
                if word in words:
                    words[word] += 1
                else:
                    words[word] = 1

                                
     # write words to whiltelist.txt
    for aWord in words:
        if words[aWord] >= 5:
            whitelist.write(aWord + '\n')
    
    whitelist.close()

if __name__=="__main__":
    write_Whitelist()



flake8_spellcheck/__init__.py
import enum
import importlib.metadata
import logging
import os
import re
import tokenize
from argparse import Namespace
from ast import AST
from itertools import chain
from pathlib import Path
from string import ascii_lowercase, ascii_uppercase, digits
from tokenize import TokenInfo
from typing import Any, FrozenSet, Iterable, Iterator, List, Optional, Tuple, Type

from flake8.options.manager import OptionManager

logger = logging.getLogger(__name__)


NOQA_REGEX = re.compile(r"#[\s]*noqa:[\s]*[\D]+[\d]+")

DICTIONARY_PATH = Path(__file__).parent
DEFAULT_DICTIONARY_NAMES = ("en_US", "python", "technical")


LintError = Tuple[int, int, str, Type["SpellCheckPlugin"]]
Position = Tuple[int, int]


class WordCase(enum.Enum):
    URL = enum.auto()
    SNAKE = enum.auto()
    CAMEL = enum.auto()


# Really simple detection function
def detect_case(word: str) -> WordCase:
    if word.startswith("http"):
        return WordCase.URL
    # ignore leading underscores when testing for snake case
    elif "_" in word.lstrip("_"):
        return WordCase.SNAKE
    elif word.isupper():
        return WordCase.SNAKE
    else:
        return WordCase.CAMEL


def parse_camel_case(name: str, position: Position) -> Iterator[Tuple[Position, str]]:
    index = position[1]
    start = index
    buffer = ""
    for c in name:
        index += 1
        if c in ascii_lowercase or c in digits or c in "'":
            buffer += c
        else:
            if buffer:
                yield (position[0], start), buffer
            if c in ascii_uppercase:
                buffer = c
                start = index - 1
            else:
                buffer = ""
                start = index

    if buffer:
        yield (position[0], start), buffer


def parse_snake_case(name: str, position: Position) -> Iterator[Tuple[Position, str]]:
    index = position[1]
    start = index
    buffer = ""
    for c in name:
        index += 1
        if c in ascii_lowercase or c in digits or c in ascii_uppercase:
            buffer += c
        else:
            if buffer:
                yield (position[0], start), buffer

            buffer = ""
            start = index

    if buffer:
        yield (position[0], start), buffer


def is_number(value: Any) -> bool:
    try:
        float(value)
    except ValueError:
        return False
    else:
        return True


def get_code(token_type: int) -> str:
    if token_type == tokenize.COMMENT:
        return "SC100"
    elif token_type == tokenize.NAME:
        return "SC200"
    else:
        raise ValueError(f"Unknown token_type {token_type}")


def find_allowlist_path(options: Namespace) -> Optional[Path]:
    if options.spellcheck_allowlist:
        if options.spellcheck_allowlist.exists():
            return options.spellcheck_allowlist
        else:
            logger.error("ERROR: Supplied allowlist file for flake8-spellcheck does not exist.")
            return None
    elif options.whitelist:
        logger.warning(
            "DEPRECATED: Support for '--whitelist' will be removed in future. Please use '--spellcheck-allowlist' instead."
        )
        whitelist_path = Path(options.whitelist)
        if whitelist_path.exists():
            return whitelist_path
        else:
            logger.error("ERROR: Supplied allowlist file for flake8-spellcheck does not exist.")
            return None

    default_allowlist_path = Path("spellcheck-allowlist.txt")
    if default_allowlist_path.exists():
        return default_allowlist_path

    default_whitelist_path = Path("whitelist.txt")
    if default_whitelist_path.exists():
        logger.warning(
            "DEPRECATED: Support for 'whitelist.txt' will be removed in future. Please use 'spellcheck-allowlist.txt' instead."
        )
        return default_whitelist_path

    return None


def flatten_dictionary_add_list(options: Namespace) -> Iterator[str]:
    for identifier in options.spellcheck_add_dictionary:
        if isinstance(identifier, str):
            yield identifier
            continue

        for _id in identifier:
            yield _id


class SpellCheckPlugin:
    name = "flake8-spellcheck"
    version = importlib.metadata.version(__name__)

    spellcheck_targets: FrozenSet[str] = frozenset()
    no_symbols: FrozenSet[str] = frozenset()
    words: FrozenSet[str] = frozenset()

    def __init__(
        self,
        tree: AST,
        filename: str = "(none)",
        file_tokens: Optional[Iterable[TokenInfo]] = None,
    ) -> None:
        if file_tokens is None:
            raise ValueError("Plugin requires file_tokens")
        else:
            self.file_tokens: Iterable[TokenInfo] = file_tokens

    @classmethod
    def load_dictionaries(cls, options: Namespace) -> Tuple[FrozenSet[str], FrozenSet[str]]:
        words = set()

        dictionary_names: Iterable[str]
        if options.dictionaries:
            logger.warning(
                "DEPRECATED: Support for '--dictionaries' will be removed in future. Use '--spellcheck-add-dictionary' instead."
            )
            dictionary_names = options.dictionaries
        else:
            if options.spellcheck_disable_default_dictionaries:
                dictionary_names = flatten_dictionary_add_list(options)
            else:
                dictionary_names = chain(
                    DEFAULT_DICTIONARY_NAMES, flatten_dictionary_add_list(options)
                )

        for dictionary_name in dictionary_names:
            dictionary_path = DICTIONARY_PATH / "{}.txt".format(dictionary_name)
            if dictionary_path.exists():
                dictionary_data = dictionary_path.read_text()
                words |= set(word.lower() for word in dictionary_data.split("\n"))
            else:
                logger.error(
                    "ERROR: Supplied built-in dictionary '{}' does not exist.".format(
                        dictionary_name
                    )
                )

        allowlist_path: Optional[Path] = find_allowlist_path(options)
        if allowlist_path:
            allowlist_data = allowlist_path.read_text()
            words |= set(w.lower() for w in allowlist_data.split("\n"))

        # Hacky way of getting dictionary with symbols stripped
        no_symbols = set()
        for w in words:
            if w.endswith("'s"):
                no_symbols.add(w.replace("'s", ""))
            else:
                no_symbols.add(w.replace("'", ""))
        return frozenset(words), frozenset(no_symbols)

    @classmethod
    def add_options(cls, parser: OptionManager) -> None:
        parser.add_option(
            "--spellcheck-allowlist",
            help="Path to text file containing a custom list of allowed words.",
            type=Path,
            parse_from_config=True,
        )
        parser.add_option(
            "--whitelist",
            help="(deprecated) Path to text file containing a custom list of allowed words. Use '--spellcheck-allowlist' instead.",
            parse_from_config=True,
        )
        parser.add_option(
            "--spellcheck-disable-default-dictionaries",
            help="Don't use the default list of built-in dictionaries. You can still use '--spellcheck-add-dictionary' to select individual built-in dictionaries.",
            action="store_true",
            parse_from_config=True,
        )
        parser.add_option(
            "--spellcheck-add-dictionary",
            help="A built-in dictionary to enable. Pass this flag multiple times to enable multiple dictionaries.",
            action="append",
            default=[],
            parse_from_config=True,
            comma_separated_list=True,
        )
        parser.add_option(
            "--dictionaries",
            help="(deprecated) A comma-separated list of built-in dictionaries to enable. Use '--spellcheck-add-dictionary' instead.",
            comma_separated_list=True,
            parse_from_config=True,
        )
        # TODO: Convert this to use action=append
        parser.add_option(
            "--spellcheck-targets",
            help="Specify the targets to spellcheck",
            default="names,comments",
            comma_separated_list=True,
            parse_from_config=True,
        )

    @classmethod
    def parse_options(cls, options: Namespace) -> None:
        cls.words, cls.no_symbols = cls.load_dictionaries(options)
        cls.spellcheck_targets = frozenset(options.spellcheck_targets)

    def _detect_errors(
        self, tokens: Iterable[Tuple[Position, str]], use_symbols: bool, token_type: int
    ) -> Iterator[LintError]:
        code = get_code(token_type)

        for position, token in tokens:
            test_token = token.lower().strip("'").strip('"')

            if use_symbols:
                valid = test_token in self.words
            else:
                valid = test_token in self.no_symbols

            # Need a way of matching words without symbols
            if not valid and not is_number(token):
                yield (
                    position[0],
                    position[1],
                    f"{code} Possibly misspelt word: '{token}'",
                    type(self),
                )

    def run(self) -> Iterator[LintError]:
        for token_info in self.file_tokens:
            yield from self._parse_token(token_info)

    def _is_valid_comment(self, token_info: tokenize.TokenInfo) -> bool:
        return (
            token_info.type == tokenize.COMMENT
            and "comments" in self.spellcheck_targets
            # Ensure comment is neither empty nor a sequence of "#" characters
            # github.com/MichaelAquilina/flake8-spellcheck/issues/34
            and token_info.string.lstrip("#").strip() != ""
            # Ignore flake8 pragma comments
            and token_info.string.lstrip("#").split()[0] != "noqa:"
        )

    def _parse_token(self, token_info: tokenize.TokenInfo) -> Iterator[LintError]:
        if token_info.type == tokenize.NAME and "names" in self.spellcheck_targets:
            value = token_info.string
        elif self._is_valid_comment(token_info):
            # strip out all `noqa: [code]` style comments so they aren't erroneously checked
            # see https://github.com/MichaelAquilina/flake8-spellcheck/issues/36 for info
            value = NOQA_REGEX.sub("", token_info.string.lstrip("#"))
        else:
            return

        tokens: List[Tuple[Position, str]] = []
        for word in value.split(" "):
            case = detect_case(word)
            if case == WordCase.URL:
                # Nothing to do here
                continue
            elif case == WordCase.SNAKE:
                tokens.extend(parse_snake_case(word, token_info.start))
            elif case == WordCase.CAMEL:
                tokens.extend(parse_camel_case(word, token_info.start))

        if token_info.type == tokenize.NAME:
            use_symbols = False
        elif token_info.type == tokenize.COMMENT:
            use_symbols = True
        else:
            return

        yield from self._detect_errors(tokens, use_symbols, token_info.type)


__all__ = ("__version__", "SpellCheckPlugin")
